### PR TITLE
refactor: using reference instead reference to unique_ptr with object

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
Many objects created and functions called by passing `const std::unique_ptr<Obj>& obj` instead directly passing `Obj& obj`

In some cases it is indeed needed, but in most cases it is just extra complexity that is better to avoid.

Motivation:
 - providing reference to object instead `unique_ptr` is giving warranty that there's no `nullptr` and no need to keep it in mind
 - value inside unique_ptr by reference can be changed externally and instead `nullptr` it can turn to real object later (or in opposite)
 - code is shorter but cleaner

Based on that  this refactoring is useful as it reduces mental load when reading or writing code.
`std::unique` should be used ONLY for owning object, but not for passing it everywhere.

## What was done?
Replaced most of usages `std::unique_ptr<Obj>& obj` to `Obj& obj`.
Btw, in several cases implementation assumes that object can be nullptr and replacement to reference is not possible.
Even using raw pointer is not possible, because the empty std::unique_ptr can be initialized later somewhere in code.
For example, in `src/init.cpp` there's called `PeerManager::make` and pass unique_ptr to the `node.llmq_ctx` that would be initialized way later.
That is out of scope this PR.
List of cases, where reference to `std::unique_ptr` stayed as they are:
 - `std::unique_ptr<LLMQContext>& llmq_ctx` in `PeerManagerImpl`, `PeerManager` and `CDSNotificationInterface`
 - `std::unique_ptr<CDeterministicMNManager>& dmnman`  in `CDSNotificationInterface`

Also `CChainState` have 3 references to `unique_ptr` that can't be replaced too:
 - `std::unique_ptr<llmq::CChainLocksHandler>& m_clhandler;`
 - `std::unique_ptr<llmq::CInstantSendManager>& m_isman;`
 - `std::unique_ptr<llmq::CQuorumBlockProcessor>& m_quorum_block_processor;`

## How Has This Been Tested?
Run unit/functional tests.

## Breaking Changes
No breaking changes, all of these changes - are internal APIs for Dash Core developers only.

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by kittywhiskers** [comment]
This PR is quite welcome, _true_ de-globalization would require non-trivial restructuring of the codebase, which is something which we cannot allocate as many resources as it would take to accomplish it in a reasonable time, so the second best thing is to lay the foundation for it and hope we get time for get to the finish line later. 

Removing unnecessary `unique_ptrs` containers for deferred dereferencing will not only make the code cleaner but further outline which parts of the codebase that actually need restructuring. 

Concept ACK

**Review by kittywhiskers — COMMENTED** [review]
There may be some instances non-const `unique_ptr` references, which could allow downstream modifications of the container's contents (such as calling `reset()`), which isn't desirable. Could as part of the review, can constness be enforced and exemptions documented?

Also, are we calling assert when we always ought to?

**Comment by PastaPastaPasta** [comment]
@knst Why is this a Draft?

**Comment by knst** [comment]
>  Why is this a Draft?

Ready now, @PastaPastaPasta

File: src/coinjoin/coinjoin.cpp [hunk 1]
```diff
@@ -457,16 +457,16 @@ void CCoinJoin::CheckDSTXes(const CBlockIndex* pindex, const llmq::CChainLocksHa
     LogPrint(BCLog::COINJOIN, "CCoinJoin::CheckDSTXes -- mapDSTX.size()=%llu\n", mapDSTX.size());
 }
 
-void CCoinJoin::UpdatedBlockTip(const CBlockIndex* pindex, const llmq::CChainLocksHandler& clhandler, const std::unique_ptr<CMasternodeSync>& mn_sync)
+void CCoinJoin::UpdatedBlockTip(const CBlockIndex* pindex, const llmq::CChainLocksHandler& clhandler, const CMasternodeSync &mn_sync)
```
Comment (UdjinM6) [inline src/coinjoin/coinjoin.cpp:460]:
```suggestion
void CCoinJoin::UpdatedBlockTip(const CBlockIndex* pindex, const llmq::CChainLocksHandler& clhandler, const CMasternodeSync& mn_sync)
```
https://github.com/dashpay/dash/blob/master/doc/developer-notes.md#coding-style-c

same for all similar changes

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge

**Review by UdjinM6 — APPROVED** [review]
one nit, LGTM otherwise

utACK

**Review by kittywhiskers — APPROVED** [review]
utACK

**Review by UdjinM6 — APPROVED** [review]
re-utACk

**Review by kittywhiskers — APPROVED** [review]
re-utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge

**Pull request merged**